### ISSUE TITLE
mount /etc/machine-id as optional default to support D-Bus

### ISSUE THREAD

**Issue by loveshack** [hpc/charliecloud#1050]
I tried to run a graphical program (paraver) and got a dbus-related error about /etc/machine-id being empty.  Mounting that into the container fixes it.

**Comment by reidpr** [comment]
Thanks @loveshack. I was not familiar with this file. Looks like it is a unique ID of the OS installation ([man page](https://man7.org/linux/man-pages/man5/machine-id.5.html), [Debian wiki](https://wiki.debian.org/MachineId)). So by bind-mounting it from the host, we declare that the host and the container are the same OS installation. That's certainly a kludge, but it could be a good kludge. Questions:

* Are you aware of any arguments that this is or is not a best practice for containers?
* `--bind /etc/machine-id` is available now as a workaround; is that sufficient for your use case?
* If we do want to initialize the file with unique but properly formatted value, it's straightforward to do this with `/dev/urandom` and POSIX tools.
* Would a FAQ entry be sufficient for your use case?

**Comment by reidpr** [comment]
I spoke with my colleague @mej about this, who's an expert on such matters. His opinions:

* There does not appear to be a consensus on whether container `/etc/machine-id` should match the host or be generated new for each container.
* It does seem that the file should contain a valid value when the container is running. On the other hand, when not running the file should be empty. (This implies that if we don't match the host, we need to keep track of the value some other way.)
* This ID is sometimes used for software licensing, which points to matching the host.
* Bind-mounting from the host seems to be a common approach (e.g. [1](https://github.com/denisbrodbeck/machineid/issues/5) [2](https://forums.docker.com/t/host-machine-id-visible-from-containers/100533/4)).

Given that, I'm inclined to proceed with the proposed solution.

**Comment by loveshack** [comment]
https://www.redhat.com/en/resources/cloud-native-container-design-whitepaper
says in support:  "Some containerized applications require the container
to be synchronized with the host on certain attributes such as time and
machine ID".  (Not that I've gone "cloud native".)

**Comment by loveshack** [comment]
In case it's still relevant, I should have said that an explicit --bind
works, of course.

[The best I came up with for generating the id with plain POSIX utils
was an od | cut | tr with appropriate args.]

**Comment by mej** [comment]
I also wanted to quickly add that the images/repositories/`rootfs` artifacts that provide the "initial state" of the filesystem for each container should *always* be empty or missing (preferably the former); this avoids the potential "worst-case" situation:  unintentionally having multiple running containers on multiple physical hosts seeing a non-unique value for something that - in theory - is guaranteed to be unique among disparate physical entities.

Imagine a distributed application that used the value of `/etc/machine-id` to choose between, say, TCP/IP sockets for remote comms vs. UNIX sockets for optimized local comms; if *all* the containers on *all* the hosts thought they were all local and tried to communicate that way...well, I suspect hilarity would ensue! 😜 

End-users can always, of course, fire upon as many of their own feet as they like, but the software should be setting them up for success, not failure.  😁

TL;DR:  I recommend that `/etc/machine-id` be (1) a zero-byte file in the directory/filesystem image and (2) `bind`-mounted from the host at runtime.

**Comment by loveshack** [comment]
That all sounds sensible, and maybe worthy input to https://wiki.debian.org/MachineId!
